File: packages/lynx-linking/android/com/sigx/linking/LinkingState.kt
package com.sigx.linking

import android.content.Intent
import android.util.Log

/**
 * Static façade between the host MainActivity and the per-LynxView
 * [LinkingPublisher] instances. The host calls [handleNewIntent] from its
 * `onCreate` (initial intent) and `onNewIntent` (warm-start) overrides;
 * publishers attached *before* the URL arrived receive it via the registered
 * listener, while publishers attached *after* read [latestURL] during their
 * init so cold-start links still seed `__globalProps.initialURL`.
 *
 * Reflectively callable by the MainActivity template — apps that don't pull
 * in `@sigx/lynx-linking` keep compiling because the template uses
 * `Class.forName("com.sigx.linking.LinkingState")` rather than a direct
 * import.
 */
object LinkingState {
    private const val TAG = "LinkingState"

    @Volatile
    var latestURL: String? = null
        private set

    private val listeners = mutableListOf<(String) -> Unit>()
    private val interceptors = mutableListOf<(String) -> Boolean>()
    private val lock = Any()

    /**
     * Forward an incoming deep-link intent from the host MainActivity. Safe
     * to call from any thread; listeners are invoked synchronously on the
     * caller's thread (publishers hop to the LynxView's thread internally
     * via the Lynx bridge).
     */
    @JvmStatic
    fun handleNewIntent(intent: Intent?) {
        val url = intent?.data?.toString() ?: return

        // Interceptors get first refusal. If one claims the URL (returns true)
        // it is considered consumed and is NOT published as a deep-link event —
        // this is how `@sigx/lynx-webauth` captures the OAuth redirect back to
        // the app's scheme without it also reaching `useLinkingNav` & friends.
        val interceptorSnapshot = synchronized(lock) { interceptors.toList() }
        for (interceptor in interceptorSnapshot) {
            val claimed =
                try {
                    interceptor(url)
                } catch (e: Throwable) {
                    Log.w(TAG, "Interceptor threw: ${e.message}")
                    false
                }
            if (claimed) return
        }

        val snapshot: List<(String) -> Unit>
        synchronized(lock) {
            latestURL = url
            snapshot = listeners.toList()
        }
        for (listener in snapshot) {
            try {
                listener(url)
            } catch (e: Throwable) {
                Log.w(TAG, "Listener threw: ${e.message}")
            }
        }
    }

    /** Register a listener; returns an unsubscribe function. */
    fun addListener(listener: (String) -> Unit): () -> Unit {
        synchronized(lock) { listeners.add(listener) }
        return {
            synchronized(lock) { listeners.remove(listener) }
        }
    }

    /**
     * Register a one-shot-style interceptor consulted *before* a deep link is
     * published. Returning `true` consumes the URL so it is NOT delivered as a
     * `Linking` 'url' event (nor stored as [latestURL]); returning `false`
     * lets it fall through to listeners as usual. Returns an unsubscribe.
     *
     * Used by `@sigx/lynx-webauth` to claim the OAuth callback redirect for the
     * duration of an auth session, so it's never mistaken for an app deep link.
     */
    fun addInterceptor(interceptor: (String) -> Boolean): () -> Unit {
        synchronized(lock) { interceptors.add(interceptor) }
        return {
            synchronized(lock) { interceptors.remove(interceptor) }
        }
    }
}

File: packages/lynx-linking/ios/LinkingState.swift
import Foundation

/// Static façade between the host AppDelegate and the per-LynxView
/// `LinkingPublisher` instances. The host calls `handleOpenURL(_:)` from its
/// `application(_:open:options:)` and universal-link continuation paths;
/// publishers attached *before* the URL arrived receive it via the posted
/// notification, while publishers attached *after* read `latestURL` during
/// their init so cold-start links still seed `__globalProps.initialURL`.
///
/// `@objc` so the AppDelegate template can find the class via
/// `NSClassFromString("LinkingState")` without compile-time dependency on
/// this package — apps that don't pull in `@sigx/lynx-linking` keep
/// compiling.
@objc public class LinkingState: NSObject {

    /// Posted by `handleOpenURL(_:)`. Active publishers observe this and
    /// forward the URL to their `LynxView`. `userInfo["url"]` is a String.
    public static let didReceiveURL = Notification.Name("SigxLynxLinkingDidReceiveURL")

    /// Most recent URL the host forwarded. Read by `LinkingPublisher.init`
    /// so a publisher attached after a cold-start deep link still sees the
    /// URL on first paint.
    public private(set) static var latestURL: String?

    private static let queue = DispatchQueue(label: "com.sigx.linking.state")

    /// Forward an incoming URL from the host AppDelegate. Safe to call from
    /// any thread; the notification dispatch hops to the main queue.
    @objc public static func handleOpenURL(_ url: URL) {
        let urlString = url.absoluteString
        queue.sync { latestURL = urlString }
        DispatchQueue.main.async {
            NotificationCenter.default.post(
                name: didReceiveURL,
                object: nil,
                userInfo: ["url": urlString]
            )
        }
    }
}
